### PR TITLE
feat: zero cooldown during chomper mode

### PR DESCRIPTION
**Pull request by plasticbugs**
This PR removes the weapon "cool down" during the alpha monster/chopper mode.

When alpha monster mode begins, we return the ball immediately to Mr. Do (this is how the arcade game works) if the ball is in "cool down" state (not bouncing, not in the process of hitting something, etc). This is Bit 6 of the ball state memory.

And then his weapon cool down is set to 1 for the duration of the mode (also how the arcade game handles the cooldown during this mode)

After the mode, the cool down will begin increasing again from 1.

### CODE REVIEW

**Comment by plasticbugs** [comment]
whoops, temporarily closing this. We don't want to return the ball to Mr. Do if it's bouncing around.

**Comment by plasticbugs** [comment]
Ok, this is ready again! Will only return the ball after the dessert is collected if the ball is in "cool down" state. We don't want to return the ball to Mr. Do if it's bouncing, hitting another enemy, etc.

File: mrdo_arcade.asm [hunk 1]
```diff
@@ -2345,11 +2345,22 @@ DEAL_WITH_BALL:
 	AND		7FH
 	OR		40H
 	LD		(IY+0), A
+
+	LD      A, ($72C3)         ; Load chomper mode flag
+	AND     0FEH               ; Mask off lowest bit (FE = 11111110)
+	CP      80H                ; Check if it's 0x80 or 0x81
```
Comment (artrag) [inline mrdo_arcade.asm:2351]: if you want to test bit 7, the correct sequence is:

LD A, ($72C3) ; Load chomper mode flag
AND $80        ; now we have 0 or 128 (i.e. bit 7)
JR    Z,.continue_normally ;  if Z the bit 7 is reset and we are i normal mode

File: mrdo_arcade.asm [hunk 2]
```diff
@@ -2345,11 +2345,22 @@ DEAL_WITH_BALL:
...
 	OR		40H
 	LD		(IY+0), A
+
+	LD      A, ($72C3)         ; Load chomper mode flag
+	AND     0FEH               ; Mask off lowest bit (FE = 11111110)
+	CP      80H                ; Check if it's 0x80 or 0x81
+	JR      NZ, .continue_normally  ; If NOT in special mode, continue normally
+	; We are in special mode
+	LD      A, 1               ; Fast cooldown during special mode
+	LD      (IY+4), A
```
Comment (artrag) [inline mrdo_arcade.asm:2355]: This sequence is better coded without using A, in this way 
LD (IY+4),1

File: mrdo_arcade.asm [hunk 3]
```diff
@@ -8333,14 +8344,52 @@ LOC_B8B1:
...
+	; Bit 6 is set to 1 when the ball is in the cooldown phase
+	; We only want to return the ball during this interval
+
+	LD      IY, $72D9         ; Load ball state pointer
+	BIT     6, (IY+0)         ; Check if bit 6 is set
+	JR      NZ, SKIP_BALL_RETURN ; Skip if not in cooldown phase
+
+  ; Return the ball
+	LD      A, 20H            ; Set only bit 5 (00100000)
+	LD      (IY+0), A         ; Store initial state
```
Comment (artrag) [inline mrdo_arcade.asm:8361]: This sequence is overwriting the whole value of (IY+0).
If you want to set bit 5 only, you can use the SET instruction and replace these two lines by

SET 5,(IY+0)

File: mrdo_arcade.asm [hunk 4]
```diff
@@ -8333,14 +8344,52 @@ LOC_B8B1:
...
+	SET     3, A              ; Set bit 3
+	LD      (IY+0), A         ; Store final state
+	LD      (IY+5), 0         ; Clear auxiliary state
+
+	CALL    PLAY_BALL_RETURN_SOUND
+	POP     IX
+
+SKIP_BALL_RETURN:
+	LD      A, 1
+	LD      (IY+4), A         ; Set cooldown to 1
```
Comment (artrag) [inline mrdo_arcade.asm:8387]: Here you can do

LD (IY+4),1

File: mrdo_arcade.asm [hunk 5]
```diff
@@ -8333,14 +8344,52 @@ LOC_B8B1:
...
+	LD      HL, 1
+	XOR     A
+	CALL    REQUEST_SIGNAL
+	LD      (IY+3), A         ; Store signal result
+
+	; Now do the ball return sequence
+	LD      A, (IY+0)
+	RES     5, A              ; Clear bit 5
+	SET     3, A              ; Set bit 3
+	LD      (IY+0), A         ; Store final state
```
Comment (artrag) [inline mrdo_arcade.asm:8379]: These 4 instructions can be replaced by
RES 5,(IY+0)
SET 3,(IY+0)

**Review by artrag — CHANGES_REQUESTED** [review]
I think you have correctly interpreted the meaning of the flag bits in $72C3, but I am not really into the game logic. 
I have suggested some changes to your code that can be made shorter and more efficient.
If you merge my comments in you base code I will download your version and merge a few changes I did to MrDo sprite management   (I want to add the death sequence, but still I need to study a bit more...)

File: mrdo_arcade.asm [hunk 6]
```diff
@@ -2343,11 +2343,22 @@ DEAL_WITH_BALL:
...
 	LD		(IY+0), A
+
+	LD      A, ($72C3)         ; Load chomper mode flag
+	AND     0FEH               ; Mask off lowest bit (FE = 11111110)
+	CP      80H                ; Check if it's 0x80 or 0x81
+	JR      NZ, .continue_normally  ; If NOT in special mode, continue normally
+	; We are in special mode
+	LD      A, 1               ; Fast cooldown during special mode
+	LD      (IY+4), A
+	JR      .play_sound
```
Comment (artrag) [inline mrdo_arcade.asm:2356]: You can avoid the jump to .play_sound if you do this:
```
	
	LD 		A,(GAMEFLAGS)
	AND		$80
	JR		Z,.normal_mode
	LD (IY+4), 0		; Fast cooldown in chomper mode
.normal_mode:	
	INC		(IY+4)		; In chomper mode set to 1 the ball cooldown counter


```

**Review by artrag — CHANGES_REQUESTED** [review]

**Pull request merged**